public/app/features/alerting/unified/rule-list/hooks/useFilteredRulesIterator.ts

```diff
@@ -0,0 +1,176 @@
+import { hashRule } from '../../utils/rule-id';
+import { isAlertingRule } from '../../utils/rules';
+
+export interface RuleWithOrigin {
+  /**
+   * Artificial frontend-only identifier for the rule.
+   * It's used as a key for the rule in the rule list to prevent key duplication
+   */
+  ruleKey: string;
+  rule: PromRuleDTO;
+  groupIdentifier: DataSourceRuleGroupIdentifier;
+}
+
+const { useLazyGroupsQuery } = prometheusApi;
+
+export function useFilteredRulesIteratorProvider() {
+  const [fetchGroups] = useLazyGroupsQuery();
+  const allExternalRulesSources = getExternalRulesSources();
+
+  /**
+   * This async generator will continue to yield rule groups and will keep fetching backend pages as long as the consumer
+   * is iterating.
+   */
+  const fetchRuleSourceGroups = useCallback(
+    async function* (ruleSource: ExternalRulesSourceIdentifier, maxGroups: number) {
+      const response = await fetchGroups({ ruleSource: { uid: ruleSource.uid }, groupLimit: maxGroups });
+
+      if (response.data?.data) {
+        yield* response.data.data.groups.map((group) => [ruleSource, group] as const);
+      }
+
+      let lastToken: string | undefined = undefined;
+      if (response.data?.data?.groupNextToken) {
+        lastToken = response.data.data.groupNextToken;
+      }
+
+      while (lastToken) {
+        const response = await fetchGroups({
+          ruleSource: { uid: ruleSource.uid },
+          groupNextToken: lastToken,
+          groupLimit: maxGroups,
+        });
+
+        if (response.data?.data) {
+          yield* response.data.data.groups.map((group) => [ruleSource, group] as const);
+        }
+
+        lastToken = response.data?.data?.groupNextToken;
+      }
+    },
+    [fetchGroups]
+  );
+
+  const getFilteredRulesIterator = (filterState: RulesFilter, groupLimit: number) => {
+    const ruleSourcesToFetchFrom = filterState.dataSourceNames.length
+      ? filterState.dataSourceNames.map((ds) => ({ name: ds, uid: getDatasourceAPIUid(ds) }))
+      : allExternalRulesSources;
+
+    // This split into the first one and the rest is only for compatibility with the merge function from ix
+    const [source, ...iterables] = ruleSourcesToFetchFrom.map((ds) => fetchRuleSourceGroups(ds, groupLimit));
+
+    return merge(source, ...iterables).pipe(
+      filter(([_, group]) => groupFilter(group, filterState)),
+      flatMap(([rulesSource, group]) => group.rules.map((rule) => [rulesSource, group, rule] as const)),
+      filter(([_, __, rule]) => ruleFilter(rule, filterState)),
+      map(([rulesSource, group, rule]) => mapRuleToRuleWithOrigin(rulesSource, group, rule))
+    );
+  };
+
+  return { getFilteredRulesIterator };
+}
+
+function mapRuleToRuleWithOrigin(
+  rulesSource: ExternalRulesSourceIdentifier,
+  group: PromRuleGroupDTO,
+  rule: PromRuleDTO
+): RuleWithOrigin {
+  const ruleKey = `${rulesSource.name}-${group.file}-${group.name}-${rule.name}-${rule.type}-${hashRule(rule)}`;
```

Comment:
nit: I think we should have a separate function to create a unique key from a rule and I don't feel like it belongs in this map function